planner/cascades/expr_iterator.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cascades
+
+import (
+	"container/list"
+)
+
+// ExprIter enumerates all the equivalent expressions in the group according to
+// the expression pattern.
+type ExprIter struct {
+	// The group and element field solely identify a group expression.
+	group   *Group
+	element *list.Element
+
+	// operand is the node of the pattern tree. The operand type of the group
+	// expression must be matched with it, otherwise the group expression is
+	// ignored during the iteration.
+	operand int
+
+	// children is used to iterate the child expressions.
+	children []*ExprIter
+}
+
+func (iter *ExprIter) Next() bool {
+	// Iterate child firstly.
+	for i := range iter.children {
+		if iter.children[i].Next() {
+			return true
+		}
+	}
+
+	// It's root node.
+	if iter.group == nil {
+		return false
+	}
+
+	// Otherwise, iterate itself to find more matched equivalent expressions.
+	for iter.element.Next(); iter.element != nil; iter.element.Next() {
+		expr := iter.element.Value.(*GroupExpr)
+		exprOperand := expr.exprNode.GetOperand()
+
+		if !iter.operand.match(exprOperand) {
+			continue
+		}
+
+		allMatched := true
+		for i := range iter.children {
+			if !iter.children[i].Reset(expr.children[i]) {
+				allMatched = false
+				break
+			}
+		}
+		if allMatched {
+			return true
+		}
+	}
+	return false
+}
+
+func (iter *ExprIter) Reset(g *Group) bool {
+	for elem := g.equivalents.Front(); elem != nil; elem.Next() {
```

Comment:
Shall we remember the first matched element of each Group in `newChildExprIter`, then we can reset the iterator to that element directly, instead of searching from the front once again?

Reply:
👍 add a `map[Operand]*list.Element` in the `Group` can help us to archive this.